### ISSUE TITLE
Crash in Sprout when running under low stress

### ISSUE THREAD

**Issue by eleanor-merry** [Metaswitch/sprout#1124]
When running Sprout under low stress, I hit this crash:

```
#0  0x00007fee97148b99 in __libc_waitpid (pid=26817, stat_loc=stat_loc@entry=0x7fee63fe3ba0, options=options@entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:40
#1  0x00007fee970ce2e2 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00000000004c5a63 in Logger::backtrace (this=0x1632290, data=<optimized out>) at /home/ubuntu/iss1090/sprout2/sprout/modules/cpp-common/src/logger.cpp:284
#3  0x000000000057779d in Log::backtrace (fmt=fmt@entry=0x6f74a8 "Signal %d caught") at /home/ubuntu/iss1090/sprout2/sprout/modules/cpp-common/src/log.cpp:185
#4  0x00000000005c1d9c in signal_handler (sig=6) at main.cpp:1019
#5  <signal handler called>
#6  0x00007fee970becc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#7  0x00007fee970c20d8 in __GI_abort () at abort.c:89
#8  0x00007fee970fb394 in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7fee97209b28 "*** Error in `%s': %s: 0x%s ***\n") at ../sysdeps/posix/libc_fatal.c:175
#9  0x00007fee9710766e in malloc_printerr (ptr=<optimized out>, str=0x7fee97209c10 "double free or corruption (!prev)", action=1) at malloc.c:4996
#10 _int_free (av=<optimized out>, p=<optimized out>, have_lock=0) at malloc.c:3840
#11 0x00000000005b9e7c in SproutletProxy::UASTsx::on_new_client_response (this=0x7fee4f118610, uac_tsx=<optimized out>, rsp=0x7fee26e78138) at sproutletproxy.cpp:549
#12 0x0000000000592c4e in BasicProxy::UACTsx::on_tsx_state (this=0x7fee03a687d0, event=0x7fee63fe6b50) at basicproxy.cpp:1904
#13 0x00000000006433c2 in tsx_set_state (tsx=0x7fee201111c8, state=PJSIP_TSX_STATE_COMPLETED, event_src_type=PJSIP_EVENT_RX_MSG, event_src=0x7fee42109f58) at ../src/pjsip/sip_transaction.c:1199
#14 0x0000000000646bf7 in tsx_on_state_proceeding_uac (tsx=0x7fee201111c8, event=0x7fee63fe6c30) at ../src/pjsip/sip_transaction.c:2908
#15 0x0000000000645fb8 in tsx_on_state_calling (tsx=0x7fee201111c8, event=0x7fee63fe6c30) at ../src/pjsip/sip_transaction.c:2491
#16 0x00000000006445d4 in pjsip_tsx_recv_msg (tsx=0x7fee201111c8, rdata=0x7fee42109f58) at ../src/pjsip/sip_transaction.c:1754
#17 0x00000000006428ee in mod_tsx_layer_on_rx_response (rdata=0x7fee42109f58) at ../src/pjsip/sip_transaction.c:863
#18 0x000000000062c1d6 in pjsip_endpt_process_rx_data (endpt=0x7fee8dd160b8, rdata=0x7fee42109f58, p=0x7fee63fe6d90, p_handled=0x0) at ../src/pjsip/sip_endpoint.c:893
#19 0x00000000005bf316 in worker_thread (p=<optimized out>) at thread_dispatcher.cpp:153
#20 0x00000000006a8e15 in thread_main (param=0x1d6c440) at ../src/pj/os_core_unix.c:523
#21 0x00007fee97c75182 in start_thread (arg=0x7fee63fe7700) at pthread_create.c:312
#22 0x00007fee9718247d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

Possibly the same cause as #1108

**Comment by JimBunch** [comment]
Marking as critical since it is service affecting. What does low stress equate to in BHCA?

**Comment by mirw** [comment]
I've been investigating this.  I can repro this every 15-20 minutes during overload conditions (~200kBHCA on an m1.small).

The stack I see is as follows.

```
#0  0x00000000000000b1 in ?? ()
#1  0x00000000005c972c in SproutletProxy::UASTsx::on_new_client_response (this=0x7fffdcd76f60, uac_tsx=<optimized out>, rsp=0x44cb0b8)
    at sproutletproxy.cpp:557
#2  0x00000000005a178e in BasicProxy::UACTsx::on_tsx_state (this=0x7fffdcd6b980, event=0x7fffbfe97b40) at basicproxy.cpp:1890
#3  0x0000000000666ebf in tsx_set_state (tsx=0x7fffdcd62fb8, state=PJSIP_TSX_STATE_COMPLETED, event_src_type=PJSIP_EVENT_RX_MSG, event_src=0x7fffaece47c8)
    at ../src/pjsip/sip_transaction.c:1199
#4  0x000000000066a84c in tsx_on_state_proceeding_uac (tsx=0x7fffdcd62fb8, event=0x7fffbfe97c20) at ../src/pjsip/sip_transaction.c:3007
#5  0x0000000000669ab5 in tsx_on_state_calling (tsx=0x7fffdcd62fb8, event=0x7fffbfe97c20) at ../src/pjsip/sip_transaction.c:2491
#6  0x00000000006680d1 in pjsip_tsx_recv_msg (tsx=0x7fffdcd62fb8, rdata=0x7fffaece47c8) at ../src/pjsip/sip_transaction.c:1754
#7  0x00000000006663eb in mod_tsx_layer_on_rx_response (rdata=0x7fffaece47c8) at ../src/pjsip/sip_transaction.c:863
#8  0x000000000064fcd3 in pjsip_endpt_process_rx_data (endpt=0x7fffebd2a0b8, rdata=0x7fffaece47c8, p=0x7fffbfe97d90, p_handled=0x0)
    at ../src/pjsip/sip_endpoint.c:893
#9  0x00000000005ceb5d in worker_thread (p=<optimized out>) at thread_dispatcher.cpp:153
#10 0x00000000006cc912 in thread_main (param=0x137d878) at ../src/pj/os_core_unix.c:523
#11 0x00007ffff5796182 in start_thread (arg=0x7fffbfe98700) at pthread_create.c:312
#12 0x00007ffff4ca347d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

The crash is at https://github.com/Metaswitch/sprout/blob/292dd26d39481f111f8161b0c364de1e9b79d7b3/sprout/sproutletproxy.cpp#L557 and from debugging it appears that upstream_sproutlet has been scribbled on or freed and reused - it consists of the text `w.cw-ngv.com/reg_data HTTP/1.1`, which looks like part of an HTTP request to Homestead.

**Comment by mirw** [comment]
The SIP response that triggered this is a `401 Unauthorized` challenge to a `REGISTER`.  The fact that this sproutlet is receiving a REGISTER response suggests that it was acting as an I-CSCF - S-CSCFs only receive REGISTER responses as a result of third-party registrations, and that was disabled on this system (and these are actually handled without sproutlets anyway).

**Comment by mirw** [comment]
I'm leaving this running under valgrind overnight to see if I can get more information from that.

**Comment by mirw** [comment]
OK, I think I've seen the issue that triggers this under valgrind, although it didn't cause a crash in this case, probably because the memory hadn't been reused in the meantime.  Accesses come in a series, as it walks through all the invalid fields in turn - the first access from one instance was:

```
==12037== Invalid read of size 8
==12037==    at 0x5C7D10: SproutletWrapper::process_actions(bool) (sproutletproxy.cpp:1634)
==12037==    by 0x5C972B: SproutletProxy::UASTsx::on_new_client_response(BasicProxy::UACTsx*, pjsip_tx_data*) (sproutletproxy.cpp:557)
==12037==    by 0x5A178D: BasicProxy::UACTsx::on_tsx_state(pjsip_event*) (basicproxy.cpp:1890)
==12037==    by 0x666EBE: tsx_set_state (sip_transaction.c:1199)
==12037==    by 0x66A84B: tsx_on_state_proceeding_uac (sip_transaction.c:3007)
==12037==    by 0x669AB4: tsx_on_state_calling (sip_transaction.c:2491)
==12037==    by 0x6680D0: pjsip_tsx_recv_msg (sip_transaction.c:1754)
==12037==    by 0x6663EA: mod_tsx_layer_on_rx_response (sip_transaction.c:863)
==12037==    by 0x64FCD2: pjsip_endpt_process_rx_data (sip_endpoint.c:893)
==12037==    by 0x5CEB5C: worker_thread(void*) (thread_dispatcher.cpp:153)
==12037==    by 0x6CC911: thread_main (os_core_unix.c:523)
==12037==    by 0x726D181: start_thread (pthread_create.c:312)
==12037==  Address 0x627878e8 is 152 bytes inside a block of size 248 free'd
==12037==    at 0x4C2C2BC: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12037==    by 0x5C997B: SproutletProxy::UASTsx::on_tsx_state(pjsip_event*) (sproutletproxy.cpp:632)
==12037==    by 0x666EBE: tsx_set_state (sip_transaction.c:1199)
==12037==    by 0x668633: send_msg_callback (sip_transaction.c:1881)
==12037==    by 0x654222: send_response_transport_cb (sip_util.c:1658)
==12037==    by 0x65441F: send_response_resolver_cb (sip_util.c:1721)
==12037==    by 0x655ABE: pjsip_resolve (sip_resolve.c:305)
==12037==    by 0x6502B7: pjsip_endpt_resolve (sip_endpoint.c:1144)
==12037==    by 0x6545AA: pjsip_endpt_send_response (sip_util.c:1772)
==12037==    by 0x668F8E: tsx_send_msg (sip_transaction.c:2126)
==12037==    by 0x669E1C: tsx_on_state_proceeding_uas (sip_transaction.c:2601)
==12037==    by 0x669B9C: tsx_on_state_trying (sip_transaction.c:2531)
```

...and the first access from another instance was:

```
==12037== Invalid read of size 8
==12037==    at 0x5C7CC7: SproutletWrapper::process_actions(bool) (stl_list.h:869)
==12037==    by 0x5C972B: SproutletProxy::UASTsx::on_new_client_response(BasicProxy::UACTsx*, pjsip_tx_data*) (sproutletproxy.cpp:557)
==12037==    by 0x5A178D: BasicProxy::UACTsx::on_tsx_state(pjsip_event*) (basicproxy.cpp:1890)
==12037==    by 0x666EBE: tsx_set_state (sip_transaction.c:1199)
==12037==    by 0x66A6F3: tsx_on_state_proceeding_uac (sip_transaction.c:2908)
==12037==    by 0x669AB4: tsx_on_state_calling (sip_transaction.c:2491)
==12037==    by 0x6680D0: pjsip_tsx_recv_msg (sip_transaction.c:1754)
==12037==    by 0x6663EA: mod_tsx_layer_on_rx_response (sip_transaction.c:863)
==12037==    by 0x64FCD2: pjsip_endpt_process_rx_data (sip_endpoint.c:893)
==12037==    by 0x5CEB5C: worker_thread(void*) (thread_dispatcher.cpp:153)
==12037==    by 0x6CC911: thread_main (os_core_unix.c:523)
==12037==    by 0x726D181: start_thread (pthread_create.c:312)
==12037==  Address 0x2a93e8c0 is 176 bytes inside a block of size 248 free'd
==12037==    at 0x4C2C2BC: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12037==    by 0x5C997B: SproutletProxy::UASTsx::on_tsx_state(pjsip_event*) (sproutletproxy.cpp:632)
==12037==    by 0x666EBE: tsx_set_state (sip_transaction.c:1199)
==12037==    by 0x668633: send_msg_callback (sip_transaction.c:1881)
==12037==    by 0x654222: send_response_transport_cb (sip_util.c:1658)
==12037==    by 0x65441F: send_response_resolver_cb (sip_util.c:1721)
==12037==    by 0x655ABE: pjsip_resolve (sip_resolve.c:305)
==12037==    by 0x6502B7: pjsip_endpt_resolve (sip_endpoint.c:1144)
==12037==    by 0x6545AA: pjsip_endpt_send_response (sip_util.c:1772)
==12037==    by 0x668F8E: tsx_send_msg (sip_transaction.c:2126)
==12037==    by 0x669E1C: tsx_on_state_proceeding_uas (sip_transaction.c:2601)
==12037==    by 0x669B9C: tsx_on_state_trying (sip_transaction.c:2531)
```

**Comment by mirw** [comment]
OK, so it looks like the SproutletWrapper was deleted when a response failed to be sent (maybe because the transport failed between receiving the original request and sending the response?)  However, we failed to record this and, when a response from the far end was received, we tried to reuse the already-deleted SproutletWrapper.

Note that all line numbers in my last few messages are against https://github.com/Metaswitch/sprout/tree/292dd26d39481f111f8161b0c364de1e9b79d7b3.

**Comment by mirw** [comment]
I think it's telling that the first access after deletion is:

```
==12037== Invalid read of size 8
==12037==    at 0x5C7D10: SproutletWrapper::process_actions(bool) (sproutletproxy.cpp:1634)
```

[sproutletproxy.cpp:1634](https://github.com/Metaswitch/sprout/blob/292dd26d39481f111f8161b0c364de1e9b79d7b3/sprout/sproutletproxy.cpp#L1634) is half-way through the `SproutletWrapper::process_actions` method, so how come it hasn't managed to try to read from the deleted SproutletWrapper earlier in the method?

I think the answer is that the SproutletWrapper was actually deleted as a result of the `tx_response` call [a few lines above](https://github.com/Metaswitch/sprout/blob/292dd26d39481f111f8161b0c364de1e9b79d7b3/sprout/sproutletproxy.cpp#L1629).  The deletion stack ends at...

```
...
==12037==    by 0x668F8E: tsx_send_msg (sip_transaction.c:2126)
==12037==    by 0x669E1C: tsx_on_state_proceeding_uas (sip_transaction.c:2601)
==12037==    by 0x669B9C: tsx_on_state_trying (sip_transaction.c:2531)
```

...but I think if we went further down the (truncated) stack we'd find that this was in turn triggered by the `tx_request` call.

In other words,

*   `SproutletWrapper::process_actions` determines that it's ready for the UAS to send the best response it received from the UAC back upstream
*   when it does so, it hits a transport error and fails - this destroys the `SproutletWrapper`
*   the stack unwinds and we're back in `SproutletWrapper::process_actions` in the context of a `SproutletWrapper` that has been deleted.

The obvious fix is to detect that we've re-entered the SproutletWrapper and not destroy it when we fail to send the response - I'll see if this flies.
